Source/NSGridView.m

```diff
@@ -0,0 +1,1280 @@
+        }
+      
+      current_x += [col trailingPadding] + [col width] + [[colWidths objectAtIndex: ci] floatValue] + _columnSpacing;
+      current_y -= [row bottomPadding] - [row height]; // - _rowSpacing; // add paddings after view...
+
+      // inc
+      i++;
+    }
+  END_FOR_IN(_cells)
+}
+
+- (instancetype) initWithFrame: (NSRect)frameRect
+{
+  self = [super initWithFrame: frameRect];
+
+  if (self != nil)
+    {
+      _rows = [[NSMutableArray alloc] init];
+      _columns = [[NSMutableArray alloc] init];
+      _cells = [[NSMutableArray alloc] init];
+    }
+
+  [self _refreshCells];
+  
+  return self;
+}
+
+- (instancetype) initWithViews: (NSArray *)rows // an array of arrays
+{
+  self = [self initWithFrame: NSZeroRect];
+
+  NSUInteger c = 0;
+  NSUInteger r = 0;
+
+  if (self != nil)
+    {
+      NSUInteger rc = [rows count], cc = 0;
+
+      if (rc > 0)
+        {
+          cc = [[rows objectAtIndex: 0] count];
+        }
+      
+      _cells = [[NSMutableArray alloc] initWithCapacity: rc * cc];
+      _rows = [[NSMutableArray alloc] initWithCapacity: rc];
+      _columns = [[NSMutableArray alloc] initWithCapacity: cc];
+
+      FOR_IN(NSArray*, row, rows)
+        {
+          NSArray *columns = [row objectAtIndex: c];
+          FOR_IN(NSArray*, column, columns)
+            {
+              NSView *v = [column objectAtIndex: c];
+              NSGridCell *cell = [[NSGridCell alloc] init];
+
+              if (r == 0)
+                {
+                  NSGridRow *gr = [[NSGridRow alloc] init];
+                  [gr setGridView: self];
+                  [_rows addObject: gr];
+                  [cell _setOwningRow: gr];
+                }
+
+              if (c == 0)
+                {
+                  NSGridColumn *gc = [[NSGridColumn alloc] init];
+                  [gc setGridView: self];
+                  [_columns addObject: gc];
+                  [cell _setOwningColumn: gc];
+                }
+              
+              [cell setContentView: v];
+              [_cells addObject: cell];
+              c++;
+            }
+          END_FOR_IN(columns);
+          r++;
+        }
+      END_FOR_IN(rows);
+      [self _refreshCells];
+    }  
+  return self;
+}
+
++ (instancetype) gridViewWithNumberOfColumns: (NSInteger)columnCount rows: (NSInteger)rowCount
+{
+  NSUInteger r = 0;
+  NSUInteger c = 0;
+  NSMutableArray *rows = [[NSMutableArray alloc] initWithCapacity: rowCount];
+  
+  for (r = 0; r < rowCount; r++)
+    {
+      NSMutableArray *col = [[NSMutableArray alloc] initWithCapacity: columnCount];
+      for (c = 0; c < columnCount; c++)
+        {
+          NSGridCell *gc = [[NSGridCell alloc] init];
+          [col addObject: gc];
+          RELEASE(gc);
+        }
+      [rows addObject: col];
+    }
+
+  return AUTORELEASE([self gridViewWithViews: rows]);
+}
+
++ (instancetype) gridViewWithViews: (NSArray *)rows
+{
+  return [[self alloc] initWithViews: rows];
```

Comment:
Here the autorelease is missing. Please remember to remove it from -gridViewWithNumberOfColumns:rows:, when you add it here.

Reply:
Done